ext/opcache/jit/zend_jit_ir.c

```diff
@@ -0,0 +1,16537 @@
+				ir_GUARD(ref, ir_CONST_ADDR(exit_addr));
+			} else {
+				ir_GUARD_NOT(ref, ir_CONST_ADDR(exit_addr));
+			}
+		} else if (smart_branch_opcode) {
+			if (opline->opcode == ZEND_IS_NOT_IDENTICAL) {
+				/* swap labels */
+				uint32_t tmp = target_label;
+				target_label = target_label2;
+				target_label2 = tmp;
+			}
+			ref = jit_IF_ex(jit, ref,
+				(smart_branch_opcode == ZEND_JMPZ || smart_branch_opcode == ZEND_JMPZ_EX) ? target_label2 : target_label);
+		}
+	}
+
+	if (smart_branch_opcode && !exit_addr) {
+		zend_basic_block *bb;
+
+		ZEND_ASSERT(jit->b >= 0);
+		bb = &jit->ssa->cfg.blocks[jit->b];
+		ZEND_ASSERT(bb->successors_count == 2);
+
+		if (bb->successors_count == 2 && bb->successors[0] == bb->successors[1]) {
+			ir_IF_TRUE(ref);
+			ir_MERGE_WITH_EMPTY_FALSE(ref);
+			_zend_jit_add_predecessor_ref(jit, bb->successors[0], jit->b, ir_END());
+		} else {
+			ZEND_ASSERT(bb->successors_count == 2);
+			_zend_jit_add_predecessor_ref(jit, bb->successors[0], jit->b, ref);
+			_zend_jit_add_predecessor_ref(jit, bb->successors[1], jit->b, ref);
+		}
+		jit->b = -1;
+	}
+
+	return 1;
+}
+
+static int zend_jit_bool_jmpznz(zend_jit_ctx *jit, const zend_op *opline, uint32_t op1_info, zend_jit_addr op1_addr, zend_jit_addr res_addr, uint32_t target_label, uint32_t target_label2, int may_throw, uint8_t branch_opcode, const void *exit_addr)
+{
+	uint32_t true_label = -1;
+	uint32_t false_label = -1;
+	bool set_bool = 0;
+	bool set_bool_not = 0;
+	bool always_true = 0, always_false = 0;
+	ir_ref ref, end_inputs = IR_UNUSED, true_inputs = IR_UNUSED, false_inputs = IR_UNUSED;
+	ir_type type = IR_UNUSED;
+
+	if (branch_opcode == ZEND_BOOL) {
+		set_bool = 1;
+	} else if (branch_opcode == ZEND_BOOL_NOT) {
+		set_bool = 1;
+		set_bool_not = 1;
+	} else if (branch_opcode == ZEND_JMPZ) {
+		true_label = target_label2;
+		false_label = target_label;
+	} else if (branch_opcode == ZEND_JMPNZ) {
+		true_label = target_label;
+		false_label = target_label2;
+	} else if (branch_opcode == ZEND_JMPZ_EX) {
+		set_bool = 1;
+		true_label = target_label2;
+		false_label = target_label;
+	} else if (branch_opcode == ZEND_JMPNZ_EX) {
+		set_bool = 1;
+		true_label = target_label;
+		false_label = target_label2;
+	} else {
+		ZEND_UNREACHABLE();
+	}
+
+	if (opline->op1_type == IS_CV && (op1_info & MAY_BE_REF)) {
+		ref = jit_ZVAL_ADDR(jit, op1_addr);
+		ref = jit_ZVAL_DEREF_ref(jit, ref);
+		op1_addr = ZEND_ADDR_REF_ZVAL(ref);
+	}
+
+	if (Z_MODE(op1_addr) == IS_CONST_ZVAL) {
+		if (zend_is_true(Z_ZV(op1_addr))) {
+			always_true = 1;
+		} else {
+			always_false = 1;
+		}
+	} else if (op1_info & (MAY_BE_UNDEF|MAY_BE_NULL|MAY_BE_FALSE|MAY_BE_TRUE)) {
+		if (!(op1_info & ((MAY_BE_UNDEF|MAY_BE_ANY)-MAY_BE_TRUE))) {
+			always_true = 1;
+		} else if (!(op1_info & (MAY_BE_ANY-(MAY_BE_NULL|MAY_BE_FALSE)))) {
+			if (opline->op1_type == IS_CV && (op1_info & MAY_BE_UNDEF)) {
+				ref = jit_ZVAL_ADDR(jit, op1_addr);
```

Comment:
I find type checking code like this difficult to read. Is `MAY_BE_ANY-MAY_BE_NULL` meant to mean "anything but null"? Maybe some additional defines or macros could help here. Some suggestions:

- `MAY_BE_ANY_BUT()` macro
- `#define MAY_BE_UNSET (MAY_BE_UNDEF|MAY_BE_NULL)`
- `#define MAY_BE_BOOL (MAY_BE_TRUE|MAY_BE_FALSE)`

Reply:
I agree about MAY_BE_BOOL, not sure about the rest.
It definitely make sense to think how to make these conditions more readable...
However, this is not a top-priority task now.